File: C++/049.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
//#include <array>
using namespace std;


class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<vector<int>, unsigned> hash2index;
        vector<int> hash; // use to conut 
        vector<vector<string>> res;
        vector<string> temp;
        map<vector<int>, unsigned>::const_iterator it;

        for(const auto &s : strs){
            hash.assign(26, 0);
            for(const auto &c : s){
                hash.at(c-'a') += 1;
            }
            if((it = hash2index.find(hash)) == hash2index.end()){
                hash2index.insert(make_pair(hash, hash2index.size()));
                temp = {s};
                res.push_back(temp);
            }
            else {
                res.at(it->second).push_back(s);
            }
        }
        return res;
    }
};

int main() {
    Solution s;
    vector<string> strs{"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<vector<string>> res = s.groupAnagrams(strs);
    for(auto v:res){
        for(auto s: v){
            cout<<s<<" ";
        }
        cout<<endl;
    }
    return 0;
}

File: 23. HashTable/049.py
from typing import List
from pprint import pprint

prime = (2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103)

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = []
        index_map = {}
        for s in strs:
            key = 1
            for c in s:
                key *= prime[ord(c)-97]
            
            if key in index_map:
                res[index_map[key]].append(s)
            else:
                index_map[key] = len(res)
                res.append([s])
            
        return res


if __name__ == "__main__":
    strs = ["eat","tea","tan","ate","nat","bat"]
    pprint(Solution().groupAnagrams(strs))